src/syntax_mapping.rs

```diff
@@ -29,204 +40,63 @@ pub enum MappingTarget<'a> {
-            (None, None) => (),
-        };
-
-        fn insert_git_config_global(mapping: &mut SyntaxMapping, config_home: impl AsRef<Path>) {
-            let git_config_path = config_home.as_ref().join("git");
-
-            mapping
-                .insert(
-                    &git_config_path.join("config").to_string_lossy(),
-                    MappingTarget::MapTo("Git Config"),
-                )
-                .ok();
-
-            mapping
-                .insert(
-                    &git_config_path.join("ignore").to_string_lossy(),
-                    MappingTarget::MapTo("Git Ignore"),
-                )
-                .ok();
-
-            mapping
-                .insert(
-                    &git_config_path.join("attributes").to_string_lossy(),
-                    MappingTarget::MapTo("Git Attributes"),
-                )
-                .ok();
-        }
+    pub fn insert(&mut self, from: &str, to: MappingTarget<'a>) -> Result<()> {
+        let matcher = make_glob_matcher(from)?;
+        self.custom_mappings.push((matcher, to));
+        Ok(())
+    }
 
-        mapping
+    /// Returns an iterator over all mappings. User-defined mappings are listed
+    /// before builtin mappings; mappings in front have higher precedence.
+    ///
+    /// Builtin mappings' `GlobMatcher`s are lazily compiled.
+    ///
+    /// Note that this function ignores builtin mappings that are invalid under
+    /// the current environment (i.e. their rules require an environment
+    /// variable that is unset).
+    pub fn all_mappings(&self) -> impl Iterator<Item = (&GlobMatcher, &MappingTarget<'a>)> {
+        self.custom_mappings()
+            .iter()
+            .map(|(matcher, target)| (matcher, target)) // as_ref
+            .chain(self.builtin_mappings())
     }
 
-    pub fn insert(&mut self, from: &str, to: MappingTarget<'a>) -> Result<()> {
-        let glob = GlobBuilder::new(from)
-            .case_insensitive(true)
-            .literal_separator(true)
-            .build()?;
-        self.mappings.push((glob.compile_matcher(), to));
-        Ok(())
+    // IMPRV: ideally `Item` should be `(&'static GlobMatcher, &'static MappingTarget<'static>)`
+    // but `Iterator::chain` (used in `SyntaxMapping::all_mappings`) asserts `Item = Self::Item`
+    // so we need a lifetime downcast, which I'm not sure how to perform
+    /// Returns an iterator over all valid builtin mappings. Mappings in front
+    /// have higher precedence.
+    ///
+    /// The `GlabMatcher`s are lazily compiled.
+    ///
+    /// If a mapping rule requires an environment variable that is unset, it
+    /// will be ignored.
+    pub fn builtin_mappings(&self) -> impl Iterator<Item = (&GlobMatcher, &MappingTarget<'a>)> {
+        BUILTIN_MAPPINGS
+            .iter()
+            .filter_map(|(matcher, target)| matcher.as_ref().map(|glob| (glob, target)))
     }
```

Comment:
Anyone familiar with lifetime downcasts?

Is this possible / easy to do?

Reply:
Okay, found a hack. `Iterator::map` seems to downcast lifetimes. Why and how I have no clue.

Reply:
I asked in the community Discord and got an explanation:

> It's not that Iterator::map is "downcasting lifetimes", it's just that impl Iterator<Item = T> is invariant in T, so you need a closure to "do" the variance

https://discord.com/channels/273534239310479360/1120124565591425034/1170543402870382653

Reply:
I asked in the community Discord and got an explanation:

> It's not that Iterator::map is "downcasting lifetimes", it's just that impl Iterator<Item = T> is invariant in T, so you need a closure to "do" the variance

https://discord.com/channels/273534239310479360/1120124565591425034/1170543402870382653